README.rst
httpcache: HTTP Caching for Python
===================================

HTTP, like all well designed standards, has multiple confusing mechanisms for
caching. httpcache is a HTTP cache that knows how to use HTTP headers and
status codes to correctly cache your HTTP traffic. It's built for use with the
excellent `Requests <https://github.com/kennethreitz/requests>`_ library,
because if you're not using Requests you're probably prepared to roll your own
caching library too.

It's gloriously easy to use. If all you want is caching in Requests, all you
need to do is plug a transport adapter into your Requests session:

.. code-block:: pycon

    >>> import requests
    >>> from httpcache import CachingHTTPAdapter
    >>> s = requests.Session()
    >>> s.mount('http://', CachingHTTPAdapter())

Away you go!

If you want more control, you can use the cache data-store itself. Store your
cache entries like this:

.. code-block:: python

    from httpcache import HTTPCache
    cache = HTTPCache(capacity=50)
    cache.store(response)

And retrieve them like this:

.. code-block:: python

    cached_response = cache.retrieve(request)

Simple.

Features
--------

- Tight integration with `Requests <https://github.com/kennethreitz/requests>`_
  data structures.
- Understands ``Expires`` and ``Cache-Control`` headers.
- Knows how to interpret ``304 Not Modified`` responses.
- Can send ``If-Modified-Since`` headers.
- Aware of HTTP verbs, e.g. ``POST``.
- RFC 2616-compliant.

Installation
------------

To install httpcache, you want to run:

.. code-block:: bash

    $ pip install httpcache

If you can't do that, and you really must have httpcache, and you can't
install ``pip``, then you can try:

.. code-block:: bash

    $ easy_install httpcache

I strongly recommend you don't do that though.

Versions
--------

httpcache supports all the versions of Python that Requests does. This means
2.6, 2.7 and 3.3. It is possible that httpcache will work on other versions of
Python but we do not test on those versions and will not support them.

Contribute
----------

Contributions are always welcome! Please abide by the following rules when
contributing:

#. Check that no-one has opened an issue already covering your bug. If you open
   a duplicate issue, the maintainer will give you a stern look.
#. Fork the `Github repository`_ and start writing your tests. If you're fixing
   a bug, I recommend writing a failing test first and working until it passes.
   If you're adding a feature, you're free to add tests after you write the
   functionality, but please test the functionality thoroughly.
#. Send a Pull Request. If I don't respond within a couple of days, please
   shout at me on Twitter or via email until I do something about it.

.. _`Github repository`: https://github.com/Lukasa/httpcache


HISTORY.rst
History
-------

0.1.3 (2013-05-19)
++++++++++++++++++

* Support Python 2.6
* Support Python 3.3

0.1.2 (2013-05-18)
++++++++++++++++++

* Fix broken setup.py.

0.1.1 (2013-05-18)
++++++++++++++++++

* Cache a broader set of response codes.
* Follow RFC 2616's recommendations about caching resources with query strings.
* Don't cache non-idempotent methods.
* Non-idempotent methods invalidate their cache resources.
* Documentation!
* Actually implement the capacity parameter.

0.1.0 (2013-05-11)
++++++++++++++++++

* Provide a Requests Transport Adapter so that caching 'Just Works'.

0.0.2 (2013-05-07)
++++++++++++++++++

* Correctly cache using the 'Expires' header.
* Correctly handle some of the functionality exposed by the 'Cache-Control' header.

0.0.1 (2013-05-05)
++++++++++++++++++

* Conception
* If-Modified-Since and 304 handling.


test_httpcache.py
# -*- coding: utf-8 -*-
"""
test-httpcache.py
~~~~~~~~~~~~~~~~~

Test cases for httpcache.
"""
import httpcache
from datetime import datetime, timedelta
import requests


class TestHTTPCache(object):
    """
    Assorted tests of the HTTPCache object.
    """
    def test_can_store_responses(self):
        resp = MockRequestsResponse()
        cache = httpcache.HTTPCache()

        rc = cache.store(resp)
        assert rc

    def test_dont_store_non_200_responses(self):
        resp = MockRequestsResponse(status_code=403)
        cache = httpcache.HTTPCache()

        rc = cache.store(resp)
        assert not rc

    def test_can_retrieve_responses(self):
        resp = MockRequestsResponse()
        cache = httpcache.HTTPCache()

        cache.store(resp)
        cached_resp = cache.handle_304(resp)

        assert resp is cached_resp

    def test_can_extract_creation_date_from_response_RFC_1123(self):
        resp = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT'})
        cache = httpcache.HTTPCache()
        dt = datetime(1994, 11, 6, 8, 49, 37)

        cache.store(resp)

        assert cache._cache[resp.url]['creation'] == dt

    def test_can_extract_creation_date_from_response_RFC_850(self):
        resp = MockRequestsResponse(headers={'Date': 'Sunday, 06-Nov-94 08:49:37 GMT'})
        cache = httpcache.HTTPCache()
        dt = datetime(1994, 11, 6, 8, 49, 37)

        cache.store(resp)

        assert cache._cache[resp.url]['creation'] == dt

    def test_can_add_if_modified_since_header(self):
        resp = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT'})
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()

        cache.store(resp)
        cache.retrieve(req)

        assert req.headers['If-Modified-Since'] == 'Sun, 06 Nov 1994 08:49:37 GMT'

    def test_we_respect_the_expires_header(self):
        resp = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT',
                                             'Expires': 'Sun, 06 Nov 2034 08:49:37 GMT'})
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()

        cache.store(resp)
        cached_resp = cache.retrieve(req)

        assert cached_resp is resp

    def test_expires_headers_invalidate(self):
        resp1 = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT',
                                              'Expires': 'Sun, 06 Nov 1994 08:49:37 GMT'})
        resp2 = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT',
                                              'Expires': 'Sun, 06 Nov 1994 08:00:00 GMT'})
        cache = httpcache.HTTPCache()

        assert not cache.store(resp1)
        assert not cache.store(resp2)

    def test_expiry_of_expires(self):
        resp = MockRequestsResponse(headers={'Date': 'Sun, 06 Nov 1994 08:49:37 GMT',
                                             'Expires': 'Sun, 04 Nov 2012 08:49:37 GMT'})
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()
        earlier = timedelta(seconds=-60)
        much_earlier = timedelta(days=-1)

        cache._cache[resp.url] = {'response': resp,
                                  'creation': datetime.utcnow() + much_earlier,
                                  'expiry': datetime.utcnow() + earlier}

        cached_resp = cache.retrieve(req)

        assert cached_resp is None
        assert len(cache._cache) == 0

    def test_we_respect_cache_control_max_age(self):
        resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()
        assert cache.store(resp)

        cached_resp = cache.retrieve(req)
        assert cached_resp is resp

    def test_we_respect_no_cache(self):
        resp = MockRequestsResponse(headers={'Cache-Control': 'no-cache'})
        cache = httpcache.HTTPCache()

        assert not cache.store(resp)

    def test_we_respect_no_store(self):
        resp = MockRequestsResponse(headers={'Cache-Control': 'no-store'})
        cache = httpcache.HTTPCache()

        assert not cache.store(resp)

    def test_we_respect_max_age_zero(self):
        resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=0'})
        cache = httpcache.HTTPCache()

        assert not cache.store(resp)

    def test_cache_is_correctly_ordered(self):
        resp1 = MockRequestsResponse()
        resp2 = MockRequestsResponse()
        resp2.url += 'abc'
        resp3 = MockRequestsResponse()
        resp3.url += 'def'
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()

        cache.store(resp1)
        cache.store(resp2)
        cache.store(resp3)
        cache.store(resp2)

        cachelist = list(cache._cache.items())
        assert cachelist[0][1]['response'] is resp1
        assert cachelist[1][1]['response'] is resp3
        assert cachelist[2][1]['response'] is resp2

        cache.handle_304(req)

        cachelist = list(cache._cache.items())
        assert cachelist[0][1]['response'] is resp3
        assert cachelist[1][1]['response'] is resp2
        assert cachelist[2][1]['response'] is resp1

    def test_do_not_cache_query_strings(self):
        resp = MockRequestsResponse()
        resp.url += '?a=b'
        cache = httpcache.HTTPCache()

        assert not cache.store(resp)

    def test_cache_query_string_if_explicitly_asked(self):
        resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
        resp.url += '?a=b'
        cache = httpcache.HTTPCache()

        assert cache.store(resp)

        cached_resp = cache.handle_304(resp)
        assert cached_resp is resp

    def test_we_dont_cache_some_methods(self):
        resp = MockRequestsResponse()
        cache = httpcache.HTTPCache()

        methods = ('POST', 'PUT', 'DELETE', 'CONNECT', 'PATCH')

        for method in methods:
            resp.request.method = method
            assert not cache.store(resp)

    def test_we_invalidate_for_some_methods(self):
        resp = MockRequestsResponse()
        cache = httpcache.HTTPCache()
        req = MockRequestsPreparedRequest()

        methods = ('POST', 'PUT', 'DELETE', 'CONNECT', 'PATCH')

        for method in methods:
            assert cache.store(resp)
            req.method = method
            assert cache.retrieve(req) is None
            assert len(cache._cache) == 0

    def test_cache_has_fixed_capacity(self):
        cache = httpcache.HTTPCache(capacity=5)

        for i in range(10):
            resp = MockRequestsResponse()
            resp.url += str(i)
            assert cache.store(resp)

        assert len(cache._cache) == 5

    def test_cache_preferentially_deletes_speculative_caching(self):
        cache = httpcache.HTTPCache(capacity=5)

        for i in range(4):
            resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
            resp.url += str(i)
            assert cache.store(resp)

        test_resp = MockRequestsResponse()
        test_resp.url += 'other'
        assert cache.store(test_resp)

        resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
        assert cache.store(resp)

        assert len(cache._cache) == 5
        assert test_resp not in [cache._cache[key] for key in list(cache._cache.keys())]

    def test_cache_will_delete_expiry_caches_if_necessary(self):
        cache = httpcache.HTTPCache(capacity=5)
        test_resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
        assert cache.store(test_resp)

        for i in range(5):
            resp = MockRequestsResponse(headers={'Cache-Control': 'max-age=3600'})
            resp.url += str(i)
            assert cache.store(resp)

        assert len(cache._cache) == 5
        assert test_resp not in [cache._cache[key] for key in list(cache._cache.keys())]


class TestCachingHTTPAdapter(object):
    """
    Tests for the caching HTTP adapter.

    NOTE: These tests currently use a fork of httpbin that has an additional
    endpoint added to it. I have an open PR that adds this endpoint, but until
    such time as it gets accepted these tests cannot be run unless you use the
    same fork on your local machine. Obtain the fork from my GitHub account.
    """
    def test_we_respect_304(self):
        s = requests.Session()
        s.mount('http://', httpcache.CachingHTTPAdapter())

        r1 = s.get('http://127.0.0.1:5000/cache')
        r2 = s.get('http://127.0.0.1:5000/cache')

        assert r1 is r2

    def test_we_respect_cache_control(self):
        s = requests.Session()
        s.mount('http://', httpcache.CachingHTTPAdapter())

        r1 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Cache-Control': 'max-age=3600'})
        r2 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Cache-Control': 'max-age=3600'})

        assert r1 is r2

    def test_we_respect_expires(self):
        s = requests.Session()
        s.mount('http://', httpcache.CachingHTTPAdapter())

        r1 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Expires': 'Sun, 06 Nov 2034 08:49:37 GMT'})
        r2 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Expires': 'Sun, 06 Nov 2034 08:49:37 GMT'})

        assert r1 is r2

    def test_we_respect_cache_control_2(self):
        s = requests.Session()
        s.mount('http://', httpcache.CachingHTTPAdapter())

        r1 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Cache-Control': 'no-cache'})
        r2 = s.get('http://127.0.0.1:5000/response-headers',
                   params={'Cache-Control': 'no-cache'})

        assert r1 is not r2


class MockRequestsResponse(object):
    """
    A specially-designed Mock object that emulates the behaviour of the
    post-v1.0.0 Requests Response object. For use with testing HTTPCache.
    """
    def __init__(self,
                 status_code=200,
                 headers={},
                 body='',
                 url='http://www.test.com/'):
        self.status_code = status_code
        self.headers = headers
        self.body = body
        self.url = url
        self.request = MockRequestsPreparedRequest(url=self.url)


class MockRequestsPreparedRequest(object):
    """
    A specially-designed Mock object that emulates the behaviour of the
    post-v1.0.0 Requests PreparedRequest object. For use with testing
    HTTPCache.
    """
    def __init__(self,
                 method='GET',
                 headers={},
                 body='',
                 url='http://www.test.com/'):
        self.method = method
        self.headers = headers
        self.body = body
        self.url = url


httpcache/utils.py
# -*- coding: utf-8 -*-
"""
utils.py
~~~~~~~~

Utility functions for use with httpcache.
"""
from datetime import datetime, timedelta

RFC_1123_DT_STR = "%a, %d %b %Y %H:%M:%S GMT"
RFC_850_DT_STR = "%A, %d-%b-%y %H:%M:%S GMT"


def parse_date_header(header):
    """
    Given a date header in the form specified by RFC 2616, return a Python
    datetime object.

    RFC 2616 specifies three possible formats for date/time headers, and
    makes it clear that all dates/times should be in UTC/GMT. That is assumed
    by this library, which simply does everything in UTC. This currently does
    not parse the C asctime() string, because that's effort.

    This function does _not_ follow Postel's Law. If a format does not strictly
    match the defined strings, this function returns None. This is considered
    'safe' behaviour.
    """
    try:
        dt = datetime.strptime(header, RFC_1123_DT_STR)
    except ValueError:
        try:
            dt = datetime.strptime(header, RFC_850_DT_STR)
        except ValueError:
            dt = None
    except TypeError:
        dt = None

    return dt


def build_date_header(dt):
    """
    Given a Python datetime object, build a Date header value according to
    RFC 2616.

    RFC 2616 specifies that the RFC 1123 form is to be preferred, so that is
    what we use.
    """
    return dt.strftime(RFC_1123_DT_STR)


def expires_from_cache_control(header, current_time):
    """
    Given a Cache-Control header, builds a Python datetime object corresponding
    to the expiry time (in UTC). This function should respect all relevant
    Cache-Control directives.

    Takes current_time as an argument to ensure that 'max-age=0' generates the
    correct behaviour without being special-cased.

    Returns None to indicate that a request must not be cached.
    """
    # Cache control header values are made of multiple comma separated fields.
    # Splitting them like this is probably a bad idea, but I'm going to roll with
    # it for now. We'll come back to it.
    fields = header.split(', ')
    duration = None

    for field in fields:
        # Right now we don't handle no-cache applied to specific fields. To be
        # as 'nice' as possible, treat any no-cache as applying to the whole
        # request. Bail early, because there's no reason to stick around.
        if field.startswith('no-cache') or field == 'no-store':
            return None

        if field.startswith('max-age'):
            _, duration = field.split('=')
            duration = int(duration)

    interval = timedelta(seconds=int(duration))

    return current_time + interval


httpcache/adapter.py
"""
adapter.py
~~~~~~~~~~

Contains an implementation of an HTTP adapter for Requests that is aware of the
cache contained in this module.
"""
from requests.adapters import HTTPAdapter
from .cache import HTTPCache


class CachingHTTPAdapter(HTTPAdapter):
    """
    A HTTP-caching-aware Transport Adapter for Python Requests.
    """
    def __init__(self, capacity=50, **kwargs):
        super(CachingHTTPAdapter, self).__init__(**kwargs)

        self.cache = HTTPCache(capacity=capacity)

    def send(self, request, **kwargs):
        """
        Sends a PreparedRequest object, respecting RFC 2616's rules about HTTP
        caching. Returns a Response object that may have been cached.
        """
        cached_resp = self.cache.retrieve(request)

        if cached_resp is not None:
            return cached_resp
        else:
            return super(CachingHTTPAdapter, self).send(request, **kwargs)

    def build_response(self, request, response):
        """
        Builds a Response object from a urllib3 response. May involve returning
        a cached Response.
        """
        resp = super(CachingHTTPAdapter, self).build_response(request,
                                                              response)

        if resp.status_code == 304:
            resp = self.cache.handle_304(resp)
        else:
            self.cache.store(resp)

        return resp


httpcache/structures.py
"""
structures.py
~~~~~~~~~~~~~

Defines structures used by the httpcache module.
"""
try:
    from collections import OrderedDict, MutableMapping
except ImportError:
    # We need to do something for earlier Python compatibility. Handle this
    # later.
    pass


class RecentOrderedDict(MutableMapping):
    """
    A custom variant of the OrderedDict that ensures that the object most
    recently inserted or retrieved from the dictionary is at the top of the
    dictionary enumeration.
    """
    def __init__(self, *args, **kwargs):
        self._data = OrderedDict(*args, **kwargs)

    def __setitem__(self, key, value):
        if key in self._data:
            del self._data[key]
        self._data[key] = value

    def __getitem__(self, key):
        value = self._data[key]
        del self._data[key]
        self._data[key] = value
        return value

    def __delitem__(self, key):
        del self._data[key]

    def __iter__(self):
        return iter(self._data)

    def __len__(self):
        return len(self._data)

    def __contains__(self, value):
        return self._data.__contains__(value)

    def items(self):
        return self._data.items()


httpcache/cache.py
# -*- coding: utf-8 -*-
"""
cache.py
~~~~~~~~

Contains the primary cache structure used in http-cache.
"""
from .structures import RecentOrderedDict
from .utils import (parse_date_header, build_date_header,
                    expires_from_cache_control, url_contains_query)
from datetime import datetime


# RFC 2616 specifies that we can cache 200 OK, 203 Non Authoritative,
# 206 Partial Content, 300 Multiple Choices, 301 Moved Permanently and
# 410 Gone responses. We don't cache 206s at the moment because we
# don't handle Range and Content-Range headers.
CACHEABLE_RCS = (200, 203, 300, 301, 410)

# Cacheable verbs.
CACHEABLE_VERBS = ('GET', 'HEAD', 'OPTIONS')

# Some verbs MUST invalidate the resource in the cache, according to RFC 2616.
# If we send one of these, or any verb we don't recognise, invalidate the
# cache entry for that URL. As it happens, these are also the cacheable
# verbs. That works out well for us.
NON_INVALIDATING_VERBS = CACHEABLE_VERBS


class HTTPCache(object):
    """
    The HTTP Cache object. Manages caching of responses according to RFC 2616,
    adding necessary headers to HTTP request objects, and returning cached
    responses based on server responses.

    This object is not expected to be used by most users. It is exposed as part
    of the public API for users who feel the need for more control. This API
    may change in a minor version increase. Be warned.

    :param capacity: (Optional) The maximum capacity of the HTTP cache.
    """
    def __init__(self, capacity=50):
        #: The maximum capacity of the HTTP cache. When this many cache entries
        #: end up in the cache, the oldest entries are removed.
        self.capacity = capacity

        #: The cache backing store. Cache entries are stored here as key-value
        #: pairs. The key is the URL used to retrieve the cached response. The
        #: value is a python dict, which stores three objects: the response
        #: (keyed off of 'response'), the retrieval or creation date (keyed off
        #: of 'creation') and the cache expiry date (keyed off of 'expiry').
        #: This last value may be None.
        self._cache = RecentOrderedDict()

    def store(self, response):
        """
        Takes an HTTP response object and stores it in the cache according to
        RFC 2616. Returns a boolean value indicating whether the response was
        cached or not.

        :param response: Requests :class:`Response <Response>` object to cache.
        """
        # Define an internal utility function.
        def date_header_or_default(header_name, default, response):
            try:
                date_header = response.headers[header_name]
            except KeyError:
                value = default
            else:
                value = parse_date_header(date_header)
            return value

        if response.status_code not in CACHEABLE_RCS:
            return False

        if response.request.method not in CACHEABLE_VERBS:
            return False

        url = response.url
        now = datetime.utcnow()

        # Get the value of the 'Date' header, if it exists. If it doesn't, just
        # use now.
        creation = date_header_or_default('Date', now, response)

        # Get the value of the 'Cache-Control' header, if it exists.
        cc = response.headers.get('Cache-Control', None)
        if cc is not None:
            expiry = expires_from_cache_control(cc, now)

            # If the above returns None, we are explicitly instructed not to
            # cache this.
            if expiry is None:
                return False

        # Get the value of the 'Expires' header, if it exists, and if we don't
        # have anything from the 'Cache-Control' header.
        if cc is None:
            expiry = date_header_or_default('Expires', None, response)

        # If the expiry date is earlier or the same as the Date header, don't
        # cache the response at all.
        if expiry is not None and expiry <= creation:
            return False

        # If there's a query portion of the url and it's a GET, don't cache
        # this unless explicitly instructed to.
        if expiry is None and response.request.method == 'GET':
            if url_contains_query(url):
                return False

        self._cache[url] = {'response': response,
                            'creation': creation,
                            'expiry': expiry}

        self.__reduce_cache_count()

        return True

    def handle_304(self, response):
        """
        Given a 304 response, retrieves the cached entry. This unconditionally
        returns the cached entry, so it can be used when the 'intelligent'
        behaviour of retrieve() is not desired.

        Returns None if there is no entry in the cache.

        :param response: The 304 response to find the cached entry for. Should be a Requests :class:`Response <Response>`.
        """
        try:
            cached_response = self._cache[response.url]['response']
        except KeyError:
            cached_response = None

        return cached_response

    def retrieve(self, request):
        """
        Retrieves a cached response if possible.

        If there is a response that can be unconditionally returned (e.g. one
        that had a Cache-Control header set), that response is returned. If
        there is one that can be conditionally returned (if a 304 is returned),
        applies an If-Modified-Since header to the request and returns None.

        :param request: The Requests :class:`PreparedRequest <PreparedRequest>` object.
        """
        return_response = None
        url = request.url

        try:
            cached_response = self._cache[url]
        except KeyError:
            return None

        if request.method not in NON_INVALIDATING_VERBS:
            del self._cache[url]
            return None

        if cached_response['expiry'] is None:
            # We have no explicit expiry time, so we weren't instructed to
            # cache. Add an 'If-Modified-Since' header.
            creation = cached_response['creation']
            header = build_date_header(creation)
            request.headers['If-Modified-Since'] = header
        else:
            # We have an explicit expiry time. If we're earlier than the expiry
            # time, return the response.
            now = datetime.utcnow()

            if now <= cached_response['expiry']:
                return_response = cached_response['response']
            else:
                del self._cache[url]

        return return_response

    def __reduce_cache_count(self):
        """
        Drops the number of entries in the cache to the capacity of the cache.

        Walks the backing RecentOrderedDict in order from oldest to youngest.
        Deletes cache entries that are either invalid or being speculatively
        cached until the number of cache entries drops to the capacity. If this
        leaves the cache above capacity, begins deleting the least-used cache
        entries that are still valid until the cache has space.
        """
        if len(self._cache) <= self.capacity:
            return

        to_delete = len(self._cache) - self.capacity
        keys = self._cache.keys()

        for key in keys:
            if self._cache[key]['expiry'] is None:
                del self._cache[key]
                to_delete -= 1

            if to_delete == 0:
                return

        keys = self._cache.keys()

        for i in range(to_delete):
            del self._cache[keys[i]]

        return
